File: semester 4/MAP/JFTS/FestivalDomain/src/main/java/ro/mpp/domain/Show.java
package ro.mpp.domain;

import lombok.Getter;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.util.List;

@Getter
public class Show extends Entity<Integer> {
    private final Timestamp date;
    private final String title;
    private final Integer soldSeats;
    private final List<Artist> performers;
    private final Venue venue;

    public Show(Integer id, Timestamp date, String title, Integer soldSeats, List<Artist> performers, Venue venue) {
        super(id);
        this.date = date;
        this.title = title;
        this.soldSeats = soldSeats;
        this.performers = performers;
        this.venue = venue;
    }

    public Show(ResultSet rs, List<Artist> performers, Venue venue) throws SQLException {
        super(rs.getInt("id"));
        this.date = rs.getTimestamp("date");
        this.title = rs.getString("title");
        this.soldSeats = rs.getInt("soldSeats");
        this.performers = performers;
        this.venue = venue;
    }

    public boolean isSoldOut() {
        return this.venue.getCapacity() <= this.soldSeats;
    }

    public Integer remainingSeats() {
        return this.venue.getCapacity() - this.soldSeats;
    }

    // this function will always buy the ticket, it won't verify if there are seats
    public Ticket sellTicket(String buyerName, Integer numberOfSeats) {
        return new Ticket(
                buyerName,
                numberOfSeats,
                Timestamp.valueOf(LocalDateTime.now()),
                this
        );
    }
}

File: semester 4/MPP/C#FTS/Model/Domain/Show.cs
using System;
using System.Data;

namespace C_FTS.Domain
{
    public class Show : Entity<int>
    {
        public string Date { get; set; }
        public string Title { get; }
        public int SoldSeats { get; private set; }
        public Venue Venue { get; }

        public Show(int id, string date, string title, int soldSeats, Venue venue)
            : base(id)
        {
            Date = date;
            Title = title ?? throw new ArgumentNullException(nameof(title));
            SoldSeats = soldSeats;
            Venue = venue ?? throw new ArgumentNullException(nameof(venue));
        }

        public Show(IDataReader reader, Venue venue)
            : base(reader.GetInt32(reader.GetOrdinal("id")))
        {
            Date = reader.GetString(reader.GetOrdinal("date"));
            Title = reader.GetString(reader.GetOrdinal("title"));
            SoldSeats = reader.GetInt32(reader.GetOrdinal("sold_seats"));
            Venue = venue ?? throw new ArgumentNullException(nameof(venue));
        }

        public bool IsSoldOut()
        {
            return Venue.Capacity <= SoldSeats;
        }

        // This function always sells the ticket without checking capacity
        public Ticket SellTicket(string buyerName, int numberOfSeats)
        {
            return new Ticket(
                buyerName,
                numberOfSeats,
                DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ss"),
                this
            );
        }

        public void AddToSoldSeats(int numberOfSoldSeats)
        {
            this.SoldSeats += numberOfSoldSeats;
        }

        public string FormattedDate => Date.Substring(0, 10);
        public string FormattedTime => Date.Substring(11, 5);
        public int RemainingSeats => Venue.Capacity - SoldSeats;
    }
}
